### ISSUE TITLE
Node reboot leaving existing pod using resources stuck with error UnexpectedAdmissionError 

### ISSUE THREAD

**Issue by liuyuan10** [kubernetes/kubernetes#125579]
### What happened?

When a node is rebooted, pod using resources allocated by device plugin will encounter UnexpectedAdmissionError error as below:

```
  Warning  UnexpectedAdmissionError  84s                kubelet            Allocate failed due to no healthy devices present; cannot allocate unhealthy devices xxx, which is unexpected
```

What makes it really bad is if it's a raw pod, it stucks in such state and never recover.

### What did you expect to happen?

The pod should be retried until device plugin is ready

### How can we reproduce it (as minimally and precisely as possible)?

1. create a pod requesting any resources allocated via a device plugin
2. reboot the node
3. observe pod failure

### Anything else we need to know?

The behavior is introduced with #116376

And there are various issues opened around kubelet restart #118559 #124345

But this issue is about node restart. When node is restarted, kubelet started to rerun existing pods in random order. So a pod can run into this issue before device plugin pod is healthy on the node

### Kubernetes version

<details>

```console
kubectl version
Client Version: v1.30.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.1
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by liuyuan10** [comment]
While this issue can be easily reproduced with node reboot, I believe it can also happen if device plugin suddenly becomes unhealthy after pod is scheduled and before kubelet allocates the resources.

The expected behavior is that the failure should be retried just like CNI failures.

**Comment by SamDecrock** [comment]
I can confirm this issue. Luckily for us the pods are managed by Deployments and they are restarted. But we're stills stuck with 'ghost' pods in an `UnexpectedAdmissionError` state.

**Comment by ffromani** [comment]
> ### What happened?
> 
> When a node is rebooted, pod using resources allocated by device plugin will encounter UnexpectedAdmissionError error as below:
> 
> ```
>   Warning  UnexpectedAdmissionError  84s                kubelet            Allocate failed due to no healthy devices present; cannot allocate unhealthy devices xxx, which is unexpected
> ```

Hi! This behavior wants to make evident and recoverable a previously hidden breakage when pod actually started, but noone allocated the device they requested: https://github.com/kubernetes/kubernetes/issues/109595

I don't know the mechanics of the device plugin, but it seems likely to me that this problem can totally happen also for gpu devices. IOW, from what I gathered it was _not_ device-specific but rather a flaw in how the devices are handled in kubelet, hence the flag.

> What makes it really bad is if it's a raw pod, it stucks in such state and never recover.

This behavior can be surprising indeed, but when pod reach terminal state the system doesn't try to recover them, this is why the platform introduce and recommends higher level controllers like deployment

> ### What did you expect to happen?
> 
> The pod should be retried until device plugin is ready

Yes, this is something we've being discussed since a while but there's unfortunately nothing concrete yet :\

**Comment by ffromani** [comment]
/triage accepted

kubelet retries are being discussed froma  while and there's general agreement it's a desirable behavior

**Comment by neolit123** [comment]
> The behavior is introduced with https://github.com/kubernetes/kubernetes/pull/116376

/assign @swatisehgal 
for feedback as the PR author.

**Comment by swatisehgal** [comment]
>  So a pod can run into this issue before device plugin pod is healthy on the node

This is correct and expected behavior. We want to strictly ensure before admitting the application pod that the device plugin has registered itself to kubelet AND healthy devices are present on the node to be allocated. Prior to https://github.com/kubernetes/kubernetes/pull/116376, we would admit the pod but then end up in runtime error in the workload as the device was never provisioned. 

As mentioned above, it is recommended to use controllers over standalone pods which would mean that we would have a successful pod eventually when the device plugin has re-registered to Kubelet.

Pods that failed can be removed using a command like below:

**Comment by swatisehgal** [comment]
I know that this might not seem ideal but if we have a scenario where application pod requesting a device is recovered before the device plugin has re-registered to Kubelet (after node reboot), we **cannot** fulfill the device requirement of that pod. I strongly believe that clear explicit failure at admission is better than allowing the workload that does not meet its resource requirement to run and failing trying to access the device it never got at a later stage.

**Comment by MaiSao** [comment]
Hi, any update in this issue, any workaround for this ?

**Comment by MaiSao** [comment]
> there is a 2-stage workaround:
> 
> 1. don't run bare pods, use deployments or other higher-order objects whose controller will take care of the pod recreation
> 2. run cleanup for pod which failed admission
> 
> a possible definitive solution can be probably done building on top of the upcoming "implicit tolerations" works [kubernetes/enhancements#5282](https://github.com/kubernetes/enhancements/issues/5282) currently being evaluted. There is no design yet.

Thanks. I still don't run bare pods, but when multi node reboot, we have a lot of 'ghost' pod for sure. I am waiting for the update.

**Comment by MaiSao** [comment]
In the cluster k8s v1.21.5, i don't see any 'ghost' pods like this. is it changed anything from ? @ffromani

**Comment by ffromani** [comment]
note these are not "ghost" pods. These we are talking about are  pods which failed admission on the kubelet, thus are in failed state. These pods don't consume compute resources on worker nodes, but they consume a modest amount of control plane resources. The resource cost of a failed pod is negligible, but when they pileup in the thousand the cumulative cost can be impactful.

The change being talked about here was introduced around k8s 1.28, so well after 1.21.

**Comment by lokanthak** [comment]
@ffromani , what is the solution for these  "UnexpectedAdmissionError"  pods these are almost 1000 pods , any bug fix or workaround to fix automatically
